tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc

```diff
@@ -85,6 +137,175 @@ bool DotGpuFusionStrategy::initFusionPattern(ShapeAnalysis& shapeAnalysis,
   return true;
 }
 
+namespace {
+
+void getDirectConsumerOps(Operation* op, DenseSet<Operation*>& consumers) {
+  consumers.clear();
+  if (op == nullptr) {
+    return;
+  }
+  SmallVector<Value> affectedValues;
+  for (auto result : op->getResults()) {
+    affectedValues.push_back(result);
+  }
+  for (auto operand : op->getOperands()) {
+    if (IsOpWriteValue(op, operand)) {
+      affectedValues.push_back(operand);
+    }
+  }
+
+  for (auto value : affectedValues) {
+    for (auto user : value.getUsers()) {
+      if (user == op) {
+        continue;
+      }
+      consumers.insert(user);
+    }
+  }
+}
+
+void getDirectProducerOpsInFusionPattern(Operation* op,
+                                         FusionPattern& fusion_pattern,
+                                         DenseSet<Operation*>& producers) {
+  producers.clear();
+
+  auto& op_list = fusion_pattern.getOpList();
+  DenseSet<Operation*> op_set(op_list.begin(), op_list.end());
+
+  int64_t num_input_operand = op->getNumOperands() - getNumResultOperands(op);
+  for (Value value : op->getOperands().take_front(num_input_operand)) {
+    auto producer = fusion_pattern.findLastWriter(value);
+    if (op_set.contains(producer)) {
+      producers.insert(producer);
+    }
+  }
+}
+
+}  // namespace
+
+bool DotGpuFusionStrategy::finalizeFusionPattern(
+    ShapeAnalysis& shapeAnalysis, FusionPattern& fusion_pattern,
+    SmallVectorImpl<Operation*>& excluded_ops) {
+  if (fusion_pattern.getFusionType() != FusionType::kDot) {
+    // None of my business.
+    return true;
+  }
+
+  // Currently, kDot fusion only support one root op. If there are many roots,
+  // find the joint paths between roots, and remain the shortest. This is to
+  // guarantee that there is only one output of the fusion. We will support
+  // multi-output fusion in the future.
```

Comment:
Am I right to think the lack of supporting for multiple roots a.t.m required DotFusion to exclude some nodes from current fusion pattern, which cause follow-up `reContractEdges`?